From cae4f984a88a63c9d6bc5e7773bf3383ac93cdaa Mon Sep 17 00:00:00 2001
From: Tommy Lin <lin.tommysc@inventec.com>
Date: Fri, 16 Sep 2022 03:01:27 +0000
Subject: [PATCH 7/7] Fix SEL log ID skip number

Symptom/Reason:
- SEL id skipped after PDU off/on

Root Cause:
- rsyslog skipped journal log when detecting timestamp backward

Solution/Change:
- [phosphor-sel-logger]
  - Undo patch of Bug651
  - Reset recordIndex when SEL clear
- [dbus-sensors]
  - Set ms_subsystem_check service launch after xyz.openbmc_project.Time.Manager
- [transformers-init]
  - Set watchdog service launch after xyz.openbmc_project.Time.Manager

%% original patch: 0007-Fix-SEL-log-ID-skip-number.patch
---
 src/sel_logger.cpp | 13 +++++++------
 1 file changed, 7 insertions(+), 6 deletions(-)

diff --git a/src/sel_logger.cpp b/src/sel_logger.cpp
index bce56b8..cfe34d8 100644
--- a/src/sel_logger.cpp
+++ b/src/sel_logger.cpp
@@ -165,8 +165,6 @@ static unsigned int initializeRecordId(void)
 }
 
 #ifdef SEL_LOGGER_CLEARS_SEL
-static unsigned int recordId = initializeRecordId();
-
 void clearSelLogFiles()
 {
     // Clear the SEL by deleting the log files
@@ -180,7 +178,7 @@ void clearSelLogFiles()
         }
     }
 
-    recordId = selInvalidRecID;
+    recordIndex = selInvalidRecID;
 
     // Reload rsyslog so it knows to start new log files
     boost::asio::io_context io;
@@ -238,6 +236,8 @@ static unsigned int getNewRecordId(void)
         }
     }
 #else
+    unsigned int recordId = recordIndex;
+
     if (++recordId >= selInvalidRecID)
     {
         recordId = 1;
@@ -364,11 +364,11 @@ static void selRedfishHooks([[maybe_unused]] const std::string& message,
 static void hupHandle(int number)
 {
     unsigned int next = getNewRecordId();
+    recordIndex = 0;
     if (getClearSelLogFiles())
     {
         delClearSelLogFiles();
     }
-    recordId = 0;
     std::cerr << "sel-logger hup handle record: " << next << ", sig: " << number << "\n";
 }
 #endif
@@ -387,6 +387,9 @@ int main(int, char*[])
     }
 #endif
 
+    // recordIndex should be initialized before SelAdd function registered
+    recordIndex = initializeRecordId();
+
     // IPMI SEL Object
     conn->request_name(ipmiSelObject);
     auto server = sdbusplus::asio::object_server(conn);
@@ -452,8 +455,6 @@ int main(int, char*[])
                                         return propertyValue;
                                     });
 
-        recordIndex = initializeRecordId();
-
 #ifdef SEL_LOGGER_CLEARS_SEL
 #ifndef SEL_LOGGER_SEND_TO_LOGGING_SERVICE
     // Clear SEL entries
-- 
2.17.1

